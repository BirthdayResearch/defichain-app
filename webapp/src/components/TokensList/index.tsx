import React, { useState, useEffect } from 'react';
import { History } from 'history';
import { I18n } from 'react-redux-i18n';
import { Helmet } from 'react-helmet';
import cloneDeep from 'lodash/cloneDeep';
import { filterByValue, getPageTitle } from '@/utils/utility';
import { TOKEN_LIST_PAGE_SIZE, DFI_SYMBOL } from '@/constants';
import WalletTokenCard from '@/components/TokenCard/WalletTokenCard';
import Pagination from '@/components/Pagination';
import Header from '@/containers/HeaderComponent';
import { IToken } from '@/utils/interfaces';

interface TokensListProps {
  tokens: IToken[];
  unit: string;
  accountTokens: IToken[];
  walletBalance: any;
  getPathAddress: (
    pagePath: string,
    symbol: string,
    hash: string,
    amount: string,
    address: string,
    isLPS: boolean
  ) => string;
  history: History;
  isLoadingTokens: boolean;
  pagePath: string;
}

const TokensList: React.FunctionComponent<TokensListProps> = (
  props: TokensListProps
) => {
  const {
    unit,
    history,
    tokens,
    walletBalance,
    accountTokens,
    getPathAddress,
    pagePath,
    isLoadingTokens,
  } = props;
  const defaultPage = 1;
  const [tableData, settableData] = useState<any>([]);
  const [currentPage, setCurrentPage] = useState<number>(defaultPage);
  const pageSize = TOKEN_LIST_PAGE_SIZE;
  const total = accountTokens.length;
  const pagesCount = Math.ceil(total / pageSize);
  const from = (currentPage - 1) * pageSize;
  const to = Math.min(total, currentPage * pageSize);

  function paginate(
    pageNumber,
    tokensList?: IToken[],
    appTokens: IToken[] = []
  ) {
    let clone = cloneDeep(tokensList || accountTokens);
    const keys = {};
    clone.forEach((t) => {
      keys[t.hash] = t.symbol;
    });
    appTokens = (appTokens || []).filter((t) => !keys[t.hash]);
    clone = [...clone, ...appTokens]
      .sort((a: IToken, b: IToken) => +a.hash - +b.hash)
      .filter((t: IToken) => t.hash !== DFI_SYMBOL);
    const tableData = clone.slice(
      (pageNumber - 1) * pageSize,
      pageNumber * pageSize
    );
    setCurrentPage(pageNumber);
    settableData(tableData);
  }

  useEffect(() => {
    const verifiedTokens = cloneDeep<IToken[]>(tokens || []).filter((t) => {
      t.amount = 0;
      return t.isDAT && !t.isLPS;
    });
    const tokensList: IToken[] = filterByValue(accountTokens, '');
    paginate(defaultPage, tokensList, verifiedTokens);
  }, [accountTokens?.length, isLoadingTokens, walletBalance, tokens?.length]);

  const handleCardClick = (symbol, hash, amount, address, isLPS) => {
    history.push(
      getPathAddress(pagePath, symbol, hash, amount, address, isLPS)
    );
  };
  return (
    <div className='content'>
      <WalletTokenCard
        handleCardClick={handleCardClick}
        token={{
          symbol: unit,
          symbolKey: unit,
          amount: walletBalance,
          hash: DFI_SYMBOL,
          address: '',
        }}
      />
      {tableData.map((token, index) => (
        <WalletTokenCard
          handleCardClick={handleCardClick}
          key={index}
          token={token}
        />
      ))}
      <Pagination
        label={I18n.t('containers.wallet.walletPage.walletPaginationRange', {
          to,
          total,
          from: from + 1,
        })}
        currentPage={currentPage}
        pagesCount={pagesCount}
        handlePageClick={paginate}
      />
    </div>
  );
};

export default TokensList;
