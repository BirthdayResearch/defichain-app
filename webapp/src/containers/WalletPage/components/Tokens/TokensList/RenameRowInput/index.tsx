import React from 'react';
import { Label, Input, InputGroupAddon, InputGroupText } from 'reactstrap';
import { I18n } from 'react-redux-i18n';

const RenameRowInput = (props) => {
  const { id, name, label, placeholder, value } = props;

  return (
    <>
      <Input
        type='text'
        name={`${name}`}
        value={value}
        id={`${id}`}
        placeholder={`${placeholder}`}
      />
      <Label for='pruneTo'>{label}</Label>
    </>
  );
};

export default RenameRowInput;
