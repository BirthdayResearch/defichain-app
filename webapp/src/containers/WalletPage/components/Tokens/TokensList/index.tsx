import React, { useState, useEffect } from 'react';
import { NavLink as RRNavLink, RouteComponentProps } from 'react-router-dom';
import { I18n } from 'react-redux-i18n';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import cloneDeep from 'lodash/cloneDeep';
import { Button, ButtonGroup, Row, Col } from 'reactstrap';
import { MdAdd } from 'react-icons/md';
import {
  fetchTokensRequest,
  fetchAccountTokensRequest,
  fetchInstantBalanceRequest,
} from '../../../reducer';
import {
  filterByValue,
  getPageTitle,
  getToken,
} from '../../../../../utils/utility';
import {
  handleAddToken,
  getWalletToken,
  getVerifiedTokens,
  updateWalletToken,
  getTokenForWalletDropDown,
  isWalletDropdown,
  getWalletPathAddress,
} from '../../../service';
import {
  WALLET_PAGE_PATH,
  TOKEN_LIST_PAGE_SIZE,
  DFI_SYMBOL,
  DEFAULT_TOKEN_VALUE,
} from '../../../../../constants';
import WalletTokenCard from '../../../../../components/TokenCard/WalletTokenCard';
import Pagination from '../../../../../components/Pagination';
import CreateOrRestoreWalletPage from '../../CreateOrRestoreWalletPage';
import Header from '../../../../HeaderComponent';
import { IToken } from '../../../../../utils/interfaces';
import WalletDropdown from '../../../../../components/walletDropdown';
import BigNumber from 'bignumber.js';
import { history } from '../../../../../utils/history';
interface WalletTokensListProps extends RouteComponentProps {
  tokens: IToken[];
  unit: string;
  accountTokens: IToken[];
  walletBalance: any;
  fetchTokensRequest: () => void;
  fetchAccountTokensRequest: () => void;
  isWalletCreatedFlag: boolean;
  openResetWalletDatModal: boolean;
  isLoadingTokens: boolean;
}

const WalletTokensList: React.FunctionComponent<WalletTokensListProps> = (
  props: WalletTokensListProps
) => {
  const {
    unit,
    isWalletCreatedFlag,
    tokens,
    isLoadingTokens,
    fetchTokensRequest,
    walletBalance,
    fetchAccountTokensRequest,
    accountTokens,
  } = props;
  const defaultPage = 1;
  const [tableData, settableData] = useState<any>([]);
  const [walletTableData, setwalletTableData] = useState<any>([]);
  const [tokenData, setTokenData] = useState<any>([]);
  const [verifiedTokens, setVerifiedTokens] = useState<any>([]);
  const [formState, setFormState] = useState<any>({
    amount1: '',
    hash1: '',
    symbol1: '',
  });
  const [currentPage, setCurrentPage] = useState<number>(defaultPage);
  const pageSize = TOKEN_LIST_PAGE_SIZE;
  const totalTokenData = [
    ...accountTokens,
    ...walletTableData,
    ...verifiedTokens,
  ];
  const total = [...new Set(totalTokenData)].length;
  const pagesCount = Math.ceil(total / pageSize);
  const from = (currentPage - 1) * pageSize;
  const to = Math.min(total, currentPage * pageSize);

  useEffect(() => {
    fetchTokensRequest();
    fetchInstantBalanceRequest();
    fetchAccountTokensRequest();
    const tokenMap = getToken(props.tokens);
    setTokenData(tokenMap);
    const walletToken = getWalletToken();
    setwalletTableData(walletToken);
  }, []);

  useEffect(() => {
    const verifiedTokens = getVerifiedTokens(tokens, accountTokens);
    setVerifiedTokens(verifiedTokens);
    const tokensList: IToken[] = filterByValue(accountTokens, '');
    if (currentPage === defaultPage) {
      paginate(
        defaultPage,
        [...tokensList, ...walletTableData],
        verifiedTokens
      );
    }
  }, [
    accountTokens?.length,
    isLoadingTokens,
    walletBalance,
    tokens?.length,
    walletTableData,
  ]);

  const paginate = (
    pageNumber,
    tokensList?: IToken[],
    appTokens: IToken[] = []
  ) => {
    let clone = cloneDeep(tokensList || totalTokenData);
    const keys = {};
    clone.forEach((t) => {
      keys[t.hash] = t.symbol;
    });
    appTokens = (appTokens || []).filter((t) => !keys[t.hash]);
    clone = [...clone, ...appTokens]
      .sort((a: IToken, b: IToken) => +a.hash - +b.hash)
      .filter((t: IToken) => t.hash != DFI_SYMBOL);

    updateWalletToken(clone);

    const tableData = clone.slice(
      (pageNumber - 1) * pageSize,
      pageNumber * pageSize
    );
    setCurrentPage(pageNumber);
    settableData(tableData);
  };

  const handleDropDown = (
    hash: string,
    field1: string,
    symbol: string,
    field2: string,
    name: string
  ) => {
    const tokenInfo = {
      symbol: symbol,
      symbolKey: symbol,
      amount: new BigNumber(DEFAULT_TOKEN_VALUE).toFixed(8),
      hash: hash,
      address: '',
      name: name,
    };
    const walletTokensList = handleAddToken(tokenInfo);
    setwalletTableData(walletTokensList);
    setFormState({
      ...formState,
      [field1]: hash,
      [field2]: symbol,
    });
  };

  const handleCardClick = (symbol, hash, amount, address, isLPS) => {
    history.push(
      getWalletPathAddress(
        WALLET_PAGE_PATH,
        symbol,
        hash,
        amount,
        address,
        isLPS
      )
    );
  };

  return isLoadingTokens ? (
    <div>{I18n.t('containers.tokens.tokensPage.loading')}</div>
  ) : (
    <>
      {!isWalletCreatedFlag ? (
        <div className='main-wrapper'>
          <CreateOrRestoreWalletPage />
        </div>
      ) : (
        <div className='main-wrapper'>
          <Helmet>
            <title>
              {getPageTitle(
                I18n.t('containers.wallet.walletPage.walletDeFiApp')
              )}
            </title>
          </Helmet>
          <Header>
            <h1>{I18n.t('containers.wallet.walletPage.balances')}</h1>
            <ButtonGroup
              disabled={!isWalletDropdown(totalTokenData, tokenData)}
            >
              <WalletDropdown
                tokenMap={getTokenForWalletDropDown(totalTokenData, tokenData)}
                name={1}
                formState={formState}
                handleDropdown={handleDropDown}
                dropdownLabel={'dropdownLabel'}
              >
                <Button
                  color='link'
                  disabled={!isWalletDropdown(totalTokenData, tokenData)}
                >
                  <MdAdd />
                  <span className='d-lg-inline'>
                    {I18n.t('containers.wallet.walletWalletsPage.addBalance')}
                  </span>
                </Button>
              </WalletDropdown>
            </ButtonGroup>
          </Header>
          <div className='content'>
            <WalletTokenCard
              handleCardClick={handleCardClick}
              token={{
                symbol: unit,
                symbolKey: unit,
                amount: walletBalance,
                hash: DFI_SYMBOL,
                address: '',
              }}
            />
            {tableData.map((token, index) => (
              <WalletTokenCard
                setwalletTableData={setwalletTableData}
                handleCardClick={handleCardClick}
                key={index}
                token={token}
              />
            ))}
            <Pagination
              label={I18n.t(
                'containers.wallet.walletPage.walletPaginationRange',
                {
                  to,
                  total,
                  from: from + 1,
                }
              )}
              currentPage={currentPage}
              pagesCount={pagesCount}
              handlePageClick={paginate}
            />
          </div>
        </div>
      )}
    </>
  );
};

const mapStateToProps = (state) => {
  const {
    wallet: {
      tokens,
      isTokensLoaded,
      isLoadingTokens,
      accountTokens,
      isAccountTokensLoaded,
      isAccountLoadingTokens,
      walletBalance,
      isWalletCreatedFlag,
    },
    settings: {
      appConfig: { unit },
    },
    popover: { openResetWalletDatModal },
  } = state;
  return {
    unit,
    tokens,
    isTokensLoaded,
    isLoadingTokens,
    accountTokens,
    isAccountTokensLoaded,
    isAccountLoadingTokens,
    walletBalance,
    isWalletCreatedFlag,
    openResetWalletDatModal,
  };
};

const mapDispatchToProps = {
  fetchTokensRequest,
  fetchAccountTokensRequest,
  fetchInstantBalanceRequest,
};

export default connect(mapStateToProps, mapDispatchToProps)(WalletTokensList);
