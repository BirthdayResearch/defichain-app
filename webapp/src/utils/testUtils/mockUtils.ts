import PersistentStore from '../persistentStore';
import * as isElectron from '../isElectron';
import * as Notifications from '../notifications';
import log from 'loglevel';
import { runSaga } from 'redux-saga';
import axios from 'axios';

const mockPersistentStore = (getData, setData) => {
  PersistentStore.get = jest.fn(() => getData);
  PersistentStore.set = jest.fn(() => setData);
  return PersistentStore;
};

const mockAxios = mockFunc => {
  axios.create = jest.fn().mockImplementation(() => ({
    post: mockFunc,
  }));
};
const dispatchedFunc = async (method, payload = {}) => {
  const dispatched = Array();
  let action = {};
  if (payload) action = { payload };
  await runSaga(
    {
      dispatch: action => dispatched.push(action),
    },
    method,
    action
  );
  return dispatched;
};

const mockisElectron = () => {
  return jest.spyOn(isElectron, 'isElectron');
};
const mockIpcRenderer = () => {
  return jest.spyOn(isElectron, 'ipcRendererFunc');
};

const spyLogger = () => {
  return jest.spyOn(log, 'error');
};

const mockNotification = () => {
  return (
    jest
      .spyOn(Notifications, 'default')
      // tslint:disable-next-line: no-empty
      .mockImplementation(() => {})
  );
};
export {
  mockPersistentStore,
  mockAxios,
  dispatchedFunc,
  mockisElectron,
  mockIpcRenderer,
  spyLogger,
  mockNotification,
};
