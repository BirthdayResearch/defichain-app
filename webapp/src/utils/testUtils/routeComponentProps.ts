import { createMemoryHistory, createLocation } from 'history';
import { match } from 'react-router';
const history = createMemoryHistory();
const path = `/route/:id`;
const match: match<{ id: string }> = {
  isExact: false,
  path,
  url: path.replace(':id', '1'),
  params: { id: '1' },
};

const location = createLocation(match.url);
delete history.location.key;
delete location.key;

export { history, match, location };
