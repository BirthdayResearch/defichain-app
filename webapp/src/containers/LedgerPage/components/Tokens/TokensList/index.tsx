import React, { useEffect } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { I18n } from 'react-redux-i18n';
import { connect } from 'react-redux';
import {
  fetchTokensRequest,
  fetchAccountTokensRequest,
  fetchInstantBalanceRequest,
} from '../../../reducer';
import { LEDGER_PATH } from '@/constants';
import { IToken } from 'src/utils/interfaces';
import TokensList from '@/components/TokensList';
import { RootState } from '@/app/rootReducer';
import { StatusLedger } from '@/typings/entities';
import { getWalletPathAddress } from '@/utils/utility';

interface LedgerTokensListProps extends RouteComponentProps {
  tokens: IToken[];
  unit: string;
  accountTokens: IToken[];
  walletBalance: any;
  fetchTokensRequest: () => void;
  fetchAccountTokensRequest: () => void;
  isWalletCreatedFlag: boolean;
  openResetWalletDatModal: boolean;
  isLoadingTokens: boolean;
  status: StatusLedger;
}

const LedgerTokensList: React.FunctionComponent<LedgerTokensListProps> = (
  props: LedgerTokensListProps
) => {
  const {
    unit,
    history,
    tokens,
    isLoadingTokens,
    fetchTokensRequest,
    walletBalance,
    fetchAccountTokensRequest,
    accountTokens,
    status,
  } = props;

  useEffect(() => {
    fetchTokensRequest();
    fetchInstantBalanceRequest();
    fetchAccountTokensRequest();
  }, []);

  return isLoadingTokens ? (
    <div>{I18n.t('containers.tokens.tokensPage.loading')}</div>
  ) : (
    <>
      {status !== 'connected' ? (
        <div className='main-wrapper'>
          <div>{I18n.t('containers.ledgerPage.tokens.notConnection')}</div>
        </div>
      ) : (
        <TokensList
          tokens={tokens}
          unit={unit}
          accountTokens={accountTokens}
          walletBalance={walletBalance}
          getPathAddress={getWalletPathAddress}
          history={history}
          isLoadingTokens={isLoadingTokens}
          pagePath={LEDGER_PATH}
        />
      )}
    </>
  );
};

const mapStateToProps = (state: RootState) => {
  const {
    wallet: {
      tokens,
      isLoadingTokens,
      accountTokens,
      isAccountTokensLoaded,
      isAccountLoadingTokens,
    },
    settings: {
      appConfig: { unit },
    },
    popover: { openResetWalletDatModal },
    ledgerWallet: {
      walletBalance,
      connect: { status }
    }
  } = state;
  return {
    unit,
    tokens,
    isLoadingTokens,
    accountTokens,
    isAccountTokensLoaded,
    isAccountLoadingTokens,
    walletBalance,
    openResetWalletDatModal,
    status,
  };
};

const mapDispatchToProps = {
  fetchTokensRequest,
  fetchAccountTokensRequest,
  fetchInstantBalanceRequest,
};

export default connect(mapStateToProps, mapDispatchToProps)(LedgerTokensList);
