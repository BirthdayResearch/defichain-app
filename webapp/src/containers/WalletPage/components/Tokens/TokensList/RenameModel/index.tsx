import React from 'react';
import { MdFileDownload } from 'react-icons/md';
import { I18n } from 'react-redux-i18n';
import {
  Modal,
  ModalBody,
  ModalFooter,
  Button,
  Col,
  FormGroup,
  Row,
  InputGroupAddon,
  InputGroupText,
  InputGroup,
} from 'reactstrap';
import RenameRowInput from '../RenameRowInput';

interface RenameModelProps {
  toggles: () => void;
  renameModalOpen: boolean;
}

const RenameModel: React.FunctionComponent<RenameModelProps> = (
  props: RenameModelProps
) => {
  const { renameModalOpen, toggles } = props;

  return (
    <Modal isOpen={renameModalOpen} centered fade={false} toggle={toggles}>
      <ModalBody className='ml-3'>
        <Row className='mb-4 mt-2'>
          <Col>{I18n.t('containers.wallet.walletPage.renameWallet')}</Col>
        </Row>
        <Row>
          <Col md='12'>
            <FormGroup className='form-label-group'>
              <RenameRowInput
                fieldName={'blockHeight'}
                label={I18n.t('containers.wallet.walletPage.walletName')}
                value={I18n.t('containers.wallet.walletPage.myHappyWallet')}
                name={'blockHeight'}
                id={'maximumAmount'}
                placeholder={'text'}
              />
            </FormGroup>
          </Col>
        </Row>
      </ModalBody>
      <ModalFooter>
        <Button
          size='sm'
          color='link'
          className='ml-auto'
          onClick={() => toggles()}
        >
          {I18n.t('alerts.cancel')}
        </Button>
        <Button color='primary'>
          <span className='d-lg-inline'>
            {I18n.t('containers.wallet.walletPage.rename')}
          </span>
        </Button>
      </ModalFooter>
    </Modal>
  );
};

export default RenameModel;
