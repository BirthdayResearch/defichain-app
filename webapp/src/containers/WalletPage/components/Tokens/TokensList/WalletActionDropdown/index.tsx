import React, { useEffect, useState } from 'react';
import { MdMoreHoriz } from 'react-icons/md';
import { I18n } from 'react-redux-i18n';
import {
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
  Dropdown,
} from 'reactstrap';
import RenameModel from '../RenameModel';

interface WalletActionDropdownProps {
  renameModalOpen: boolean;
  handleRenameButtonClick: () => void;
}

const WalletActionDropdown: React.FunctionComponent<WalletActionDropdownProps> = (
  props: WalletActionDropdownProps
) => {
  const [dropdownOpen, setDropdownOpen] = useState(false);
  const toggle = () => setDropdownOpen((prevState) => !prevState);
  const [modal, setModal] = useState(true);
  const { renameModalOpen, handleRenameButtonClick } = props;

  const toggles = () => {
    const isOpen = !renameModalOpen;
    setModal(!isOpen);
    handleRenameButtonClick();
  };

  return (
    <>
      <Dropdown direction='left' isOpen={dropdownOpen} toggle={toggle}>
        <DropdownToggle
          tag='span'
          data-toggle='dropdown'
          aria-expanded={dropdownOpen}
        >
          <MdMoreHoriz />
        </DropdownToggle>
        <DropdownMenu>
          <DropdownItem>
            {I18n.t('containers.wallet.walletPage.backup')}
          </DropdownItem>
          <DropdownItem onClick={toggles}>
            {I18n.t('containers.wallet.walletPage.rename')}
          </DropdownItem>
        </DropdownMenu>
      </Dropdown>
      <RenameModel toggles={toggles} renameModalOpen={renameModalOpen} />
    </>
  );
};

export default WalletActionDropdown;
