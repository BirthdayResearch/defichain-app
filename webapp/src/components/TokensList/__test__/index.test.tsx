import React from 'react';
import TokensList from '../index';
import Enzyme from '../../../utils/testUtils/enzyme';
import * as routeComponentProps from '../../../utils/testUtils/routeComponentProps';

describe('WalletTokensList component', () => {
  it('should check for snapshot', () => {
    const wrapper = Enzyme.mount(
      <TokensList
        history={routeComponentProps.history}
        tokens={[]}
        unit={''}
        accountTokens={[]}
        walletBalance={0}
        getPathAddress={(pagePath, symbol, hash, amount, address, isLPS) => ''}
        pagePath={'/'}
        isLoadingTokens={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
