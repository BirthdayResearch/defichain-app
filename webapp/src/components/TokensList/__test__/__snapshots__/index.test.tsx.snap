// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletTokensList component should check for snapshot 1`] = `
<TokensList
  accountTokens={Array []}
  getPathAddress={[Function]}
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  isLoadingTokens={false}
  pagePath="/"
  tokens={Array []}
  unit=""
  walletBalance={0}
>
  <div
    className="content"
  >
    <WalletTokenCard
      handleCardClick={[Function]}
      token={
        Object {
          "address": "",
          "amount": 0,
          "hash": "0",
          "symbol": "",
          "symbolKey": "",
        }
      }
    >
      <Card
        onClick={[Function]}
        tag="div"
      >
        <div
          className="card"
          onClick={[Function]}
        >
          <CardBody
            tag="div"
          >
            <div
              className="card-body"
            >
              <Row
                className="align-items-center"
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <div
                  className="align-items-center row"
                >
                  <Col
                    md="6"
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    <div
                      className="col-md-6"
                    >
                      <div
                        className="d-flex align-items-center justify-content-start"
                      >
                        <div>
                          <TokenAvatar
                            symbol=""
                            textSizeRatio={2}
                          >
                            <ForwardRef(withAvatarConfig)
                              maxInitials={2}
                              name=""
                              round={true}
                              size="24px"
                              src=""
                              textSizeRatio={2}
                            >
                              <Avatar
                                avatarRedirectUrl={null}
                                cache={
                                  Cache {
                                    "cachePrefix": "react-avatar/",
                                    "sourceSize": 20,
                                    "sourceTTL": 604800000,
                                  }
                                }
                                className=""
                                colors={
                                  Array [
                                    "#d73d32",
                                    "#7e3794",
                                    "#4285f4",
                                    "#67ae3f",
                                    "#d61a7f",
                                    "#ff4080",
                                  ]
                                }
                                fgColor="#FFF"
                                initials={[Function]}
                                maxInitials={2}
                                name=""
                                round={true}
                                size="24px"
                                src=""
                                textMarginRatio={0.15}
                                textSizeRatio={2}
                                unstyled={false}
                              >
                                <div
                                  className=" sb-avatar"
                                  style={
                                    Object {
                                      "borderRadius": "100%",
                                      "display": "inline-block",
                                      "fontFamily": "Helvetica, Arial, sans-serif",
                                      "height": "24px",
                                      "verticalAlign": "middle",
                                      "width": "24px",
                                    }
                                  }
                                >
                                  <div
                                    className=" sb-avatar__text"
                                    style={
                                      Object {
                                        "background": undefined,
                                        "borderRadius": "100%",
                                        "color": "#FFF",
                                        "height": "24px",
                                        "lineHeight": "initial",
                                        "textAlign": "center",
                                        "width": "24px",
                                      }
                                    }
                                  >
                                    <div
                                      style={
                                        Object {
                                          "display": "table",
                                          "height": "100%",
                                          "tableLayout": "fixed",
                                          "width": "100%",
                                        }
                                      }
                                    >
                                      <span
                                        style={
                                          Object {
                                            "display": "table-cell",
                                            "fontSize": "100%",
                                            "verticalAlign": "middle",
                                            "whiteSpace": "nowrap",
                                          }
                                        }
                                      >
                                        <span
                                          key="24px"
                                        />
                                      </span>
                                    </div>
                                  </div>
                                </div>
                              </Avatar>
                            </ForwardRef(withAvatarConfig)>
                          </TokenAvatar>
                        </div>
                        <div
                          className="ml-4"
                        >
                          <div>
                            <b />
                          </div>
                          <div />
                        </div>
                      </div>
                    </div>
                  </Col>
                  <Col
                    md="6"
                    tag="div"
                    widths={
                      Array [
                        "xs",
                        "sm",
                        "md",
                        "lg",
                        "xl",
                      ]
                    }
                  >
                    <div
                      className="col-md-6"
                    >
                      <div
                        className="undefined justify-content-end"
                      >
                        <b
                          className="text-dark"
                        >
                          <NumberMask
                            value="0.00000000"
                          >
                            <NumberFormat
                              allowEmptyFormatting={false}
                              allowLeadingZeros={false}
                              allowNegative={true}
                              decimalSeparator="."
                              defaultValue={0}
                              displayType="text"
                              fixedDecimalScale={false}
                              isAllowed={[Function]}
                              isNumericString={true}
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onMouseUp={[Function]}
                              onValueChange={[Function]}
                              prefix=""
                              suffix=""
                              thousandSeparator={true}
                              thousandsGroupStyle="thousand"
                              type="text"
                              value="0.00000000"
                            >
                              <span>
                                0.00000000
                              </span>
                            </NumberFormat>
                          </NumberMask>
                        </b>
                        <span
                          className="ml-2"
                        />
                      </div>
                    </div>
                  </Col>
                </div>
              </Row>
            </div>
          </CardBody>
        </div>
      </Card>
    </WalletTokenCard>
    <PaginationComponent
      currentPage={1}
      handlePageClick={[Function]}
      label="1 – 0 of 0 wallets"
      pagesCount={0}
    >
      <div
        className="d-flex justify-content-between align-items-center mt-3"
      >
        <div>
          1 – 0 of 0 wallets
        </div>
        <Pagination
          aria-label="pagination"
          listTag="ul"
          tag="nav"
        >
          <nav
            aria-label="pagination"
            className=""
          >
            <ul
              className="pagination"
            >
              <PaginationItem
                disabled={true}
                tag="li"
              >
                <li
                  className="page-item disabled"
                >
                  <PaginationLink
                    first={true}
                    onClick={[Function]}
                    tag="a"
                  >
                    <button
                      aria-label="First"
                      className="page-link"
                      onClick={[Function]}
                    >
                      <span
                        aria-hidden="true"
                        key="caret"
                      >
                        <MdFirstPage>
                          <IconBase
                            attr={
                              Object {
                                "viewBox": "0 0 24 24",
                              }
                            }
                          >
                            <svg
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              strokeWidth="0"
                              style={
                                Object {
                                  "color": undefined,
                                }
                              }
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                                key="0"
                              />
                            </svg>
                          </IconBase>
                        </MdFirstPage>
                      </span>
                      <span
                        className="sr-only"
                        key="sr"
                      >
                        First
                      </span>
                    </button>
                  </PaginationLink>
                </li>
              </PaginationItem>
              <PaginationItem
                disabled={true}
                tag="li"
              >
                <li
                  className="page-item disabled"
                >
                  <PaginationLink
                    onClick={[Function]}
                    previous={true}
                    tag="a"
                  >
                    <button
                      aria-label="Previous"
                      className="page-link"
                      onClick={[Function]}
                    >
                      <span
                        aria-hidden="true"
                        key="caret"
                      >
                        <MdChevronLeft>
                          <IconBase
                            attr={
                              Object {
                                "viewBox": "0 0 24 24",
                              }
                            }
                          >
                            <svg
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              strokeWidth="0"
                              style={
                                Object {
                                  "color": undefined,
                                }
                              }
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                                key="0"
                              />
                            </svg>
                          </IconBase>
                        </MdChevronLeft>
                      </span>
                      <span
                        className="sr-only"
                        key="sr"
                      >
                        Previous
                      </span>
                    </button>
                  </PaginationLink>
                </li>
              </PaginationItem>
              <PaginationItem
                disabled={true}
                tag="li"
              >
                <li
                  className="page-item disabled"
                >
                  <PaginationLink
                    next={true}
                    onClick={[Function]}
                    tag="a"
                  >
                    <button
                      aria-label="Next"
                      className="page-link"
                      onClick={[Function]}
                    >
                      <span
                        aria-hidden="true"
                        key="caret"
                      >
                        <MdChevronRight>
                          <IconBase
                            attr={
                              Object {
                                "viewBox": "0 0 24 24",
                              }
                            }
                          >
                            <svg
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              strokeWidth="0"
                              style={
                                Object {
                                  "color": undefined,
                                }
                              }
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                                key="0"
                              />
                            </svg>
                          </IconBase>
                        </MdChevronRight>
                      </span>
                      <span
                        className="sr-only"
                        key="sr"
                      >
                        Next
                      </span>
                    </button>
                  </PaginationLink>
                </li>
              </PaginationItem>
              <PaginationItem
                disabled={true}
                tag="li"
              >
                <li
                  className="page-item disabled"
                >
                  <PaginationLink
                    last={true}
                    onClick={[Function]}
                    tag="a"
                  >
                    <button
                      aria-label="Last"
                      className="page-link"
                      onClick={[Function]}
                    >
                      <span
                        aria-hidden="true"
                        key="caret"
                      >
                        <MdLastPage>
                          <IconBase
                            attr={
                              Object {
                                "viewBox": "0 0 24 24",
                              }
                            }
                          >
                            <svg
                              fill="currentColor"
                              height="1em"
                              stroke="currentColor"
                              strokeWidth="0"
                              style={
                                Object {
                                  "color": undefined,
                                }
                              }
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                                key="0"
                              />
                            </svg>
                          </IconBase>
                        </MdLastPage>
                      </span>
                      <span
                        className="sr-only"
                        key="sr"
                      >
                        Last
                      </span>
                    </button>
                  </PaginationLink>
                </li>
              </PaginationItem>
            </ul>
          </nav>
        </Pagination>
      </div>
    </PaginationComponent>
  </div>
</TokensList>
`;
